src/optimizers/optimizer.cpp

```diff
@@ -318,94 +274,154 @@ void optimizer::add_to_gradient(const AbsDistMat& gradient,
     throw lbann_exception(err.str());
   }
   if (scale != DataType(0)) {
-    El::Axpy(scale, gradient, *m_gradient);
-    m_cpu_gradient_is_nonzero = true;
+    if (m_cudnn == nullptr) {
+      El::Axpy(scale, gradient, *m_gradient);
+    } else {
+      #ifndef LBANN_HAS_CUDNN
+      throw lbann_exception("optimizer: cuDNN not detected");
+      #else
+      cudnn::matrix gradient_d(m_cudnn,
+                               gradient.LocalHeight(),
+                               gradient.LocalWidth());
+      m_cudnn->broadcast_to_gpus(gradient_d.get_data(),
+                                 gradient.LockedMatrix());
+      add_to_gradient(gradient_d, scale);
+      #endif // LBANN_HAS_CUDNN
+    }
   }
 }
-
-void optimizer::stage_gradient_for_accumulation(const AbsDistMat& gradient,
-                                                DataType scale) {
```

Comment:
I switched the function order so that the CPU and GPU implementations of `add_to_gradient` and `stage_gradient_for_accumulation`/`add_to_gradient_staging` would be next to each other. Sorry if it makes it harder to diff the new and old versions.